Diffing for mismatches in variant and record declarations

Octachron opened the pull request:
This PR applies the recently added diffing mechanism to the error messages for mismatches in record and variant declarations in order to obtain friendly error messages when someone adds, changes or deletes a few fields or constructors in a type declaration and forget to update the corresponding interface file or a type declaration re-export.

This PR is the result of a discussion with @dra27 after #10073 where we ended up moving a constructor to avoid an unfriendly error message:

>```
>File "src/unix/lwt_unix.cppo.mli", lines 1020-1030, characters 0-13:
>Error: This variant or record definition does not match that of type
>         Unix.socket_bool_option
>       Constructors number 4 have different names, SO_REUSEPORT and SO_KEEPALIVE.
>```

Here the real issue here was the recently added `SO_REUSEPORT` constructor. Moving the new constructor to the end of type declaration solved this specific issue

>```
>File "src/unix/lwt_unix.cppo.mli", lines 1020-1030, characters 0-13:
>Error: This variant or record definition does not match that of type
>         Unix.socket_bool_option
>       The constructor SO_REUSEPORT is only present in the original definition.
>```

But with the new diffing mechanism, we have the opportunity to improve the error message in the former case too. With this PR, the error message in presence of a single erroneous addition or deletion is simplified to
```ocaml
module M: sig
  type t = A | New | B | C
end = struct
  type t = A | B | C
end
```
>```
>       ...
>       Type declarations do not match:
>         type t = A | B | C
>       is not included in
>         type t = A | New | B | C
>       A constructor, New, is missing in the first declaration.
>```

from

>```
>       ...
>       Type declarations do not match:
>         type t = A | B | C
>       is not included in
>         type t = A | New | B | C
>       Constructors number 2 have different names, B and New.
>```

The case where the error could be interpreted as a single name and/or type change is not altered.

For more involved errors, this PR reuses the scheme used for functor applications:
![err_msg](https://user-images.githubusercontent.com/7689388/115529596-541cd500-a293-11eb-8082-c1e424181e6c.png)


This is slightly less colorful than the functor case, since this PR does not alter the printing of type declaration in Oprint.

## Implementation

The implementation is quite straightforward this time: the comparison functions for single constructor and field in `Includecore` are mostly moved around. The full variant and record declaration comparison functions are rewired to use the new `Diffing` module rather than doing the analysis on their own. Finally, the last change in `Includecore` are printing functions.

Octachron committed:
diffing: define few common functions

Octachron committed:
diffing for record definition

Octachron committed:
diffing for variant definitions

Octachron committed:
more tests

Octachron committed:
update Changes

Octachron committed:
diffing: adjust weight for fields and constructors

gasche commented on testsuite/tests/typing-misc/records.ml:
```diff
@@ -264,5 +265,6 @@ Line 1, characters 0-30:
 Error: This variant or record definition does not match that of type d
-       Fields number 1 have different names, x and y.
+       1. Fields have different names, x and y.
+       2. Fields have different names, y and x.
```
(Just skimming.) This output looks confusing to me, because it has two lines that appear to say the same thing oddly differently. (And I think they do say the same thing, oddly differently.) Vague ideas:
- use an ordering to avoid duplicating those messages
- integrate permutations in the diffing algorithm (for edit distance it is common to do it only for swaps of two consecutive fields, but here I would rather consider an arbitrary move of one element as an operation).

gasche replied:
Note: it may be possible to reconstruct arbitrary-moves after the fact, by fusing a deletion and an insertion together when their endpoint are compatible. This means that the edit distance computed is different (moves cost 2 instead of 1 as we could expect), but the user message is nice. I'm not sure if the distance difference is a big deal.

gasche replied:
There should be an example in the testsuite for a larger displacement, for example

```
{a; b; c; d; e}
   {b; c; d; e; a}
```

Drup replied:
Do you consider not computing permutations a blocker for this ? I definitely think we should try to compute permutations at some point, but it's quite independent from the current patch, and would also apply to functor applications.

gasche replied:
I consider that on this example there is a clear regression from a better error message to a worse error message. I expect that this kind of situation is easy for beginners to run into, so I would be in favor of fixing this before we merge. I think that there are easy fixes -- in particular, recognizing displacements on the final proposed edition (which corresponds to minimal edits when giving displacements a cost of 2) sounds rather easy to implement. (Trying to modify the edit-distance algorithm to provide optimal suggestions for cost-1 displacements is of unknown difficulty, so maybe that should not be a requirement.)

Octachron replied:
The new version of this PR refines the error message into:

> Error: This variant or record definition does not match that of type d
       Fields x and y have been swapped.

gasche commented on testsuite/tests/typing-misc/variant.ml:
```diff
@@ -119,7 +120,8 @@ Line 1, characters 0-35:
 Error: This variant or record definition does not match that of type d
-       Constructors number 1 have different names, X and Y.
+       1. Constructors have different names, X and Y.
+       2. Constructors have different names, Y and X.
```
(same as above)

gasche commented on testsuite/tests/typing-modules/records_errors_test.ml:
```diff
@@ -134,5 +144,143 @@ Error: Signature mismatch:
+       Type declarations do not match:
+         type t = { a : unit; b : unit; beta : unit; c : unit; d : unit; }
+       is not included in
+         type t = { a : unit; b : unit; c : unit; d : unit; }
```
The "modules do not match" and "type declarations do not match" are painfully redundant. I understand that this is partly an orthogonal issue to your change, but are there plans to solve that part as well?

Octachron replied:
This is more a general question on how to print the error tree. Currently, we are emitting a full message at every step from the root of the tree to the leaf on focus. This leads to repetitions since the context at step `k` does not change much from the context at step `k+1`.  I am planing to propose another strategy (print a full context at most once), but that will be a separate PR.

gasche replied:
Sure, this should be handled separately.

gasche reviewed (dismissed):
I haven't looked at the implementation yet, but I'm not fond of the redundancy in the error messages in case a field is reordered/moved/displaced, and I think it should be fixed.

Drup commented:
I'm planing to review this (since I haven't touched the code :D), but I think it's better to wait until we are all happy with the resulting error messages. 

Personally, I think this is already quite an improvement, so I'm rather happy with it. Most of the resulting pain points are rather caused by repetitions in the module error traces, which are being worked in other PRs, or examples where computing permutations would yield better messages, which should be worked on separately.

Octachron committed:
swaps and moves

Octachron commented:
I have added a new refinement pass that identifies situations when a field or constructor has the right type and name at the wrong position. More precisely, the new module `Diffing_with_keys` recognizes swaps between two fields or constructors, e.g.
```ocaml
module M : sig
  type t = { a:int; e:int; c:int; d:int; b:int }
end = struct
  type t = { alpha:int; b:int; c:int; d:int; e:int }
end
```
>```
>       1. Fields have different names, alpha and a.
>       2<->5. Fields b and e have been swapped.
>```
and change of position:
```ocaml
module M: sig
  type t = { a:int; b:int; c:int; d:int; e:int; f:float }
end =
struct
  type t = { b:int; c:int; d:int; e:int; a:int; f:int }
end
```
>```
>       1->5. Field a has been moved from position 1 to 5.
>       6. Fields do not match:
>         f : int;
>       is not compatible with:
>         f : float;
>       The types are not equal.
>```

Along the way, I have fixed a bug in presence of existentially quantified type in inline record.
Typically, in
```ocaml
module Not_a_swap: sig
  type t = A: { x:'a; a:'a; b:'b; y:'b} -> t
end = struct
  type t = A: { y:'a; a:'a; b:'b; x:'b} -> t
end
```
the previous version of the code accepted to alpha-convert `'a` and `'b`. This bug is now gone.

gasche commented on typing/includecore.ml:
```diff
@@ -184,10 +180,11 @@ type type_mismatch =
   | Variance
   | Record_mismatch of record_mismatch
-  | Variant_mismatch of variant_mismatch
+  | Variant_mismatch of variant_change list
```
Nitpick:
- you could define the two `*_change` types first, then all `*_mismatch` types
- it would look more regular with `type variant_mismatch = variant_change list`

gasche commented on typing/includecore.ml:
```diff
@@ -197,22 +194,51 @@ let report_label_mismatch first second ppf err =
+  | Move {name; got; expected } ->
+      Format.fprintf ppf
+        "@[<2>%aField %s has been moved@ from@ position %d@ to %d.@]"
+        prefix x name expected got
```
Why is it that `Change` and `Move` messages are boxed, but not any other?

Octachron replied:
The message was more verbose at some point, and I added break hints. I will update the other messages with break hints.

gasche commented on typing/includecore.ml:
```diff
@@ -197,22 +194,51 @@ let report_label_mismatch first second ppf err =
         (String.capitalize_ascii  (choose ord first second))
         (choose_other ord first second)
 
+let pp_record_diff first second prefix decl ppf (x : record_change) =
```
It is fairly tempting to turn several of these arguments (at least `first`, `second`, `prefix`) into keywords arguments. What's holding you back? (Invasiveness of the change along the call stack?)

gasche commented on typing/includecore.ml:
```diff
@@ -294,121 +334,113 @@ let report_type_mismatch first second decl ppf err =
+    if not (equal ~loc env params1 params2 l r) then
+      let patch = diffing loc env params1 params2 l r in
+      Some (Record_mismatch (Label_mismatch patch))
+    else
```
I think it would be nicer if `diffing` would call `equal` and return an option, and then have no other call to `equal` outside the function. Morally the fact that we don't use Diffing to check equality is an optimization, so it should not incur duplication of code/logic ni the rest of the code.

Here this would be done by calling `compare` first, and then comparing representations only on `None`. This would also make the code more consistent with the variant case.

gasche commented on utils/diffing.ml:
```diff
@@ -368,3 +368,34 @@ let variadic_diff ~weight ~test ~(update:_ update) state line column =
+    | Insert _ -> 10
+    | Delete _ -> 10
+    | Change _ -> 10
+    | Keep _ -> 0
```
this last function was meant to be dedented

gasche commented on utils/diffing.mli:
```diff
@@ -110,3 +111,19 @@ val variadic_diff :
   'state -> 'l array -> 'r array -> ('l, 'r, 'eq, 'diff) patch
+
+
+val default_weight : _ change -> int
```
is this function actually used?

Octachron replied:
Not anymore, because the weights have been locally specialized to account for the `Diffing_with_keys` variant (in a commit anterior to `Diffing_with_keys`). The function itself might be useful in term of documentation.

gasche commented on typing/includecore.ml:
```diff
@@ -427,6 +459,131 @@ let compare_records_with_representation ~loc env params1 params2 n
+    | Diffing.Keep _ -> 0
+    | Diffing.Change (_,_,Diffing_with_keys.Name t) ->
+        if t.types_match then 10 else 15
+    | Diffing.Change _ -> 10
```
This function is a duplicate from the other `weight` function. And `default_weight` appears unused. Something is wrong here.

Octachron replied:
Indeed, this is the `default_weight` function for `Diffing_with_keys`, which is currently missing.

gasche commented on utils/diffing_with_keys.mli:
```diff
@@ -0,0 +1,54 @@
+  | Swap of { pos: int * int; first: string; last: string }
+  | Move of {name:string; got:int; expected:int}
+  | Insert of {pos:int; insert:'a}
+  | Delete of {pos:int; delete:'a}
```
the spacing for `Swap` is oddly inconsistent

gasche reviewed (commented):
This looks nice overall; I haven't looked at the implementation of `Diffing_with_keys` yet.

gasche commented:
I started reading `diffing_with_keys.ml` and ended up with several optional changes idea, the first of which is https://github.com/Octachron/ocaml/pull/2 .

gasche commented:
I proposed a further simplification of the Diffing codebase in https://github.com/Octachron/ocaml/pull/3 . If @Octachron agrees that this is the way to go, there is some work left for him to apply the same functorization to Diffing_with_keys, which may be part of this PR or another one.

Drup commented on utils/diffing.mli:
```diff
@@ -110,3 +111,19 @@ val variadic_diff :
+val prefix: Format.formatter -> (int * change_kind) -> unit
+val style: change_kind -> Misc.Color.style list
+
+(** Enriched analysis in presence of keys *)
```
I suppose this shouldn't be here.

Drup commented on utils/diffing_with_keys.ml:
```diff
@@ -0,0 +1,143 @@
+            else None
+        | None -> Some (Change reason)
+  in
+  List.filter_map filter patch
```
Can you discuss the algorithm here a bit ? 

It seems to me you are implicitly building the graph of position changes where each edge represents a move from position i to position i'. The way you are doing it is a bit ... diluted ? Since you specialized the 2-cycle for swap, and build 2 different maps. If you were doing it explicitly, you would immediately get the refined patch, including arbitrary permutations by finding the cycles.

Octachron replied:
The current algorithm is essentially detecting the 2-cycles of the graph of name changes in the patch (with pairs of addition and deletion considered as a cycle going through a special 𝜔 node). Since any edge of a 2-cycles fully identifies the 2-cycle it might belong to, the algorithm compute a map of partial cycles and update it . Switching to an explicit graph would clearly the way to compute larger cycles.

Drup commented on utils/diffing_with_keys.ml:
```diff
@@ -0,0 +1,143 @@
+let with_pos l = List.mapi (fun n x -> n+1,x) l
+let pos (x,_) = x
+let data (_,x) = x
+let mk_pos pos data = pos, data
```
At this point, might as well make it a record. :)

Octachron replied:
The record was a bit more noisy outside of the module, but it might be better to have a record rather than the ghost of a record.

Octachron replied:
The record has been restored.

Drup reviewed (commented):
I like the new swap/move detection very much, it gives very good messages.

I made some comments on the diffing implementation. At this point, I'm not completely convinced by the algorithm used. Morally, we are going to end up with 3 kinds of diffable things: 
1. Lists (functor/functions). Insert/Delete/Change. Swap and Moves are not free, but too complicated to compute.
2. Dictionaries (signatures, objects, polyvars). (Insert)/Delete/Change. Swaps and Moves are free.
3. Ordered dictionaries/key-ed lists (record/variants decls). Insert/Delete/Change/Swap/Move

The reason the initial PR resulted in slightly non-intuitive messages is that it considered declarations as lists who happens to have keys (3 as special case of 1), which is consistent with the compiler's view. Users, on the other, consider them as dictionaries which happens to have an order (3 as special case of 2). Maybe it would be better to implement the algorithm for string-keyed dictionaries with the full panel (I/D/C/S/M), which we can reuse for lot's of things later anyway.

This might be a bit premature and overkill.

Octachron committed:
Functorized diffing (with improved documentation) (#4)

gasche reviewed (approved):
I have now reviewed the implementation of the Diffing_with_keys module as well, so I am entirely in favor of approving the PR. Thanks!

gasche commented on Changes:
```diff
@@ -382,6 +382,11 @@ OCaml 4.13.0
 
+- #?????: Improve error messages for mismatched record and variant
+  definitions.
+  (Florian Angeletti, review by ???)
```
This needs an update.

Octachron replied:
I have updated the entry along with the pending record change.

Octachron committed:
diffing_with_keys: switch to a record

Octachron committed:
update Changes for #10361: add reviewers
